// exiftool-vendored/dist/DateTime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var luxon_1 = require("luxon");
var ExifDate_1 = require("./ExifDate");
var ExifDateTime_1 = require("./ExifDateTime");
var ExifTime_1 = require("./ExifTime");
function validDateTime(dt) {
    return dt != null && dt.isValid;
}
exports.validDateTime = validDateTime;
function diffMs(a, b) {
    return a.getTime() - b.getTime();
}
exports.diffMs = diffMs;
exports.MinuteMs = 60 * 1000;
exports.HourMs = 60 * exports.MinuteMs;
exports.DayMs = 24 * exports.HourMs;
function isDateOrTime(o) {
    return (o instanceof ExifDateTime_1.ExifDateTime ||
        o instanceof ExifDate_1.ExifDate ||
        o instanceof ExifTime_1.ExifTime ||
        o instanceof luxon_1.DateTime);
}
exports.isDateOrTime = isDateOrTime;
function toExifString(d) {
    if (d instanceof luxon_1.DateTime) {
        return dateTimeToExif(d);
    }
    else {
        return d.toExifString();
    }
}
exports.toExifString = toExifString;
function dateTimeToExif(d) {
    return d.toFormat("y:LL:dd HH:mm:ss.u");
}
exports.dateTimeToExif = dateTimeToExif;
