// exiftool-vendored/dist/ExifDate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var luxon_1 = require("luxon");
var DateTime_1 = require("./DateTime");
var Maybe_1 = require("./Maybe");
var String_1 = require("./String");
/**
 * Encodes an ExifDate
 */
var ExifDate = /** @class */ (function () {
    function ExifDate(year, // full year (probably 2019-ish, but maybe Japanese 30-ish). See https://ericasadun.com/2018/12/25/iso-8601-yyyy-yyyy-and-why-your-year-may-be-wrong/
    month, // 1-12, (no crazy 0-11 nonsense from Date!)
    day // 1-31
    ) {
        this.year = year;
        this.month = month;
        this.day = day;
    }
    ExifDate.fromISO = function (text) {
        return this.fromDateTime(luxon_1.DateTime.fromISO(text));
    };
    ExifDate.fromPatterns = function (text, fmts) {
        var _this = this;
        if (String_1.blank(text))
            return;
        text = String_1.toS(text).trim();
        return Maybe_1.first(fmts, function (fmt) {
            return Maybe_1.map(luxon_1.DateTime.fromFormat(text, fmt), function (dt) { return _this.fromDateTime(dt); });
        });
    };
    // These are all formats I've seen in the wild from exiftool's output.
    // More iterations might make sense, like "d MMM, y" or "MMM d, y", but I
    // want to be constrained in what I consider a valid date to lessen the
    // chance of misinterpreting a given value.
    ExifDate.fromExifStrict = function (text) {
        return this.fromPatterns(text, ["y:MM:dd", "y-MM-dd", "y:M:d"]);
    };
    ExifDate.fromExifLoose = function (text) {
        return this.fromPatterns(text, ["MMM d y", "MMMM d y"]);
    };
    ExifDate.fromEXIF = function (text) {
        var _this = this;
        return Maybe_1.firstDefinedThunk([
            function () { return _this.fromExifStrict(text); },
            function () { return _this.fromExifLoose(text); },
            function () { return _this.fromISO(text); }
        ]);
    };
    ExifDate.fromDateTime = function (dt) {
        return DateTime_1.validDateTime(dt)
            ? new ExifDate(dt.year, dt.month, dt.day)
            : undefined;
    };
    ExifDate.prototype.toDate = function () {
        return new Date(this.year, this.month - 1, this.day);
    };
    ExifDate.prototype.toISOString = function () {
        return this.toString("-");
    };
    ExifDate.prototype.toExifString = function () {
        return this.toString(":");
    };
    ExifDate.prototype.toString = function (sep) {
        if (sep === void 0) { sep = "-"; }
        return "" + this.year + sep + String_1.pad2(this.month, this.day).join(sep);
    };
    return ExifDate;
}());
exports.ExifDate = ExifDate;
