// exiftool-vendored/dist/ExifTime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var luxon_1 = require("luxon");
var Maybe_1 = require("./Maybe");
var String_1 = require("./String");
/**
 * Encodes an ExifTime (which may not have a timezone offset)
 */
var ExifTime = /** @class */ (function () {
    function ExifTime(hour, minute, second, millisecond) {
        var _this = this;
        this.hour = hour;
        this.minute = minute;
        this.second = second;
        this.millisecond = millisecond;
        this.toString = function () {
            return String_1.pad2(_this.hour, _this.minute, _this.second).join(":") + _this.subsec();
        };
        this.toISOString = this.toString;
        this.toExifString = this.toString;
    }
    ExifTime.fromEXIF = function (text) {
        var _this = this;
        if (String_1.blank(text))
            return;
        text = String_1.toS(text).trim();
        return Maybe_1.first(["HH:mm:ss.uZZ", "HH:mm:ssZZ", "HH:mm:ss.u", "HH:mm:ss"], function (fmt) { return Maybe_1.map(luxon_1.DateTime.fromFormat(text, fmt), function (dt) { return _this.fromDateTime(dt); }); });
    };
    ExifTime.fromDateTime = function (dt) {
        return dt == null || !dt.isValid
            ? undefined
            : new ExifTime(dt.hour, dt.minute, dt.second, dt.millisecond);
    };
    Object.defineProperty(ExifTime.prototype, "millis", {
        get: function () {
            return this.millisecond;
        },
        enumerable: true,
        configurable: true
    });
    ExifTime.prototype.subsec = function () {
        return this.millisecond == null || this.millisecond === 0
            ? ""
            : "." + String_1.pad3(this.millisecond);
    };
    return ExifTime;
}());
exports.ExifTime = ExifTime;
